### PR TITLE
[inductor] Rewrite multi-consumer F.pad as torch.cat for zero-copy

### PR DESCRIPTION
**Pull request by tianrengao**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #177216
*  #175729

Rewrite constant_pad_nd as ConcatKernel([x, fill_tensor]) when the pad
input has multiple consumers. This avoids duplicate computation by
leveraging ConcatKernel's realize-into-slice zero-copy mechanism.

Only supports right-pad on a single dimension. Single-consumer pads
are unchanged (default masked Pointwise is optimal).

Fixes: vllm-project/vllm#24917

## Summary

Fixes a performance issue where `torch.cat` and `F.pad` (constant_pad_nd)
generate redundant computation when their inputs have multiple consumers.

### vLLM MoE (mul + pad + addmm) https://github.com/vllm-project/vllm/pull/25477

```python
mul_result = x * scale                            # [tokens, 2880]
padded = F.pad(mul_result, [0, 192])              # [tokens, 3072]
mm_result = torch.addmm(bias, mul_result, weight) # also uses mul_result
```

mul_result has two consumers. constant_pad_nd lowering inlines mul via pointwise_cat with tl.where masking, but mul_result must also be realized for addmm — causing mul to be computed twice.

## Solution

Add _pad_as_cat() path in constant_pad_nd lowering: for right-pad on a single dimension with multi-consumer input, rewrite as ConcatKernel.create([x, fill_tensor], dim). This lets F.pad benefit from ConcatKernel's realize-into-slice zero-copy mechanism.

Also relaxes an assert in ConcatKernel.create() (ir.py) to allow calls from non-aten.cat contexts.

Only supports right-pad, single dimension. Single-consumer pads unchanged.

## Codegen: vLLM MoE pattern (mul + pad + addmm)
Pattern: mul [4096, 2880] + F.pad [4096, 3072] + addmm (GPT-OSS MoE, vllm#24917)

### BEFORE:
```python
buf1 = empty_strided_cuda((4096, 2880), ...)          # mul buffer
triton_poi_fused_mul_0.run(x, scale, buf1)             # Kernel 1: mul → buf1
extern_kernels.addmm(bias, buf1, weight, out=buf2)     # addmm reads buf1
del buf1
buf0 = empty_strided_cuda((4096, 3072), ...)           # padded buffer
triton_poi_fused_pad_mul_1.run(x, scale, buf0)         # Kernel 2: mul RECOMPUTED + tl.where
```

### AFTER:
```python
buf0 = empty_strided_cuda((4096, 3072), (3072, 1), ...)        # padded buffer (pre-alloc)
buf_mul = reinterpret_tensor(buf0, (4096, 2880), (3072, 1), 0) # slice alias
triton_poi_fused_mul_0.run(x, scale, buf_mul)                   # Kernel 1: mul → slice
buf_pad = reinterpret_tensor(buf0, (4096, 192), (3072, 1), 2880)
triton_poi_fused_zeros_1.run(buf_pad)                            # Kernel 2: fill zeros
extern_kernels.addmm(bias, buf_mul, weight, out=buf2)           # addmm reads same slice
```

Codegen:
  BEFORE: pointwise_cat inlines mul → mul computed 2x, 3 buffers, tl.where masking
  AFTER:  ConcatKernel zero-copy → mul computed 1x, 2 buffers, reinterpret_tensor

Perf (NVIDIA H100, bf16):
  BEFORE: 0.122 ms/iter
  AFTER:  0.085 ms/iter  (**+30% speedup**)

### E2E vLLM Serving Benchmark

**Model:** GPT-OSS 20B (MXFP4, 32 experts, hidden_size=2880→2944 padded)
**Hardware:** 2× H100 TP=2 · **vLLM** v0.15.2 · **PyTorch** 2.12.0a0+git38fbfb4
**Workload:** 500 prompts, input_len=1024, output_len=128, request_rate=inf (saturated)

| Metric | Inductor (no fix) | Inductor (with fix) | Δ |
|--------|------------------:|--------------------:|:-:|
| Request throughput (req/s) | 46.33 | 46.96 | **+1.34%** |
| Output throughput (tok/s) | 5930.82 | 6010.53 | **+1.34%** |
| Mean TTFT (ms) | 4675.75 | 4578.40 | **−2.08%** |
| Mean TPOT (ms) | 45.40 | 45.05 | **−0.76%** |
| P99 TPOT (ms) | 63.72 | 63.08 | **−1.01%** |

### Inductor Model Performance Dashboard
Improved across all the models on the inductor dashboard

Screenshot 2026-03-24 at 1 51 46 PM

## What This Does NOT Cover
- Left-pad or multi-dimension padding
- RMSNorm reduction + pad full fusion (needs scheduler changes)
- Single-consumer pad (already optimal with masked Pointwise)

## Test Plan
Added unit tests

Verified with repro scripts on NVIDIA B200:

**vLLM MoE pattern:**

| Variant | Buffers Before | Buffers After | Correctness |
|---------|---------------|---------------|-------------|
| F.pad + addmm (multi-consumer) | 3 | **2** | PASS |
| torch.cat + addmm (multi-consumer) | 3 | **2** | PASS |
| torch.cat single consumer | 1 | **1 (unchanged)** | PASS |
| F.pad single consumer | 1 | **1 (unchanged)** | PASS |

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos

### CODE REVIEW

File: torch/_inductor/lowering.py [hunk 1]
```diff
@@ -4735,6 +4735,55 @@ def _padding_can_be_fused():
...
+
+    # Single-consumer pads are already optimal with masked Pointwise.
+    pad_node = V.current_node
+    if pad_node is None:
+        return None
+    input_node = pad_node.args[0]
+    if not isinstance(input_node, torch.fx.Node):
+        return None
+    if len(input_node.users) <= 1:
+        return None
```
Comment (eellison) [inline torch/_inductor/lowering.py]: Why not just let cat handle this ? even with single consumer, there are cases where we'd want to memory plan.

File: torch/_inductor/lowering.py [hunk 2]
```diff
@@ -4735,6 +4735,55 @@ def _padding_can_be_fused():
...
+    pad_shape[pad_dim] = pad_amount
+    dtype = x.get_dtype()
+    device = x.get_device()
+    fill_value_typed = dtype_to_type(dtype)(fill_value)
+    pad_tensor = tensor_constructor(fill_value_typed)(
+        pad_shape, dtype=dtype, device=device
+    )
+
+    counters["inductor"]["pad_as_cat"] += 1
+    return TensorBox(ir.ConcatKernel.create([x, pad_tensor], pad_dim))
```
Comment (eellison) [inline torch/_inductor/lowering.py]: I was expecting this to call into to `cat` lowering, not `ir.ConcatKernel` directly

**Review by eellison — COMMENTED** [review]
Two comments. the `cat` lowering already has heuristics for when to memory plan vs fuse. can we reuse that ?

File: torch/_inductor/lowering.py (hunk 2 above)
Comment (tianrengao) [reply torch/_inductor/lowering.py]: Fixed, `pad_as_cat` now calls `cat` instead of ir.ConcatKernel.create()

To make this work, extended any_input_has_multi_consumers() in cat() to handle the pad-as-cat context. For unrealized pointwise inputs with multiple consumers, pointwise_cat would inline and recompute them, so we prefer ConcatKernel

File: torch/_inductor/lowering.py (hunk 1 above)
Comment (tianrengao) [reply torch/_inductor/lowering.py]: Done. Removed the single-consumer guard and now delegate to cat() lowering

**Comment by tianrengao** [comment]
@claude review this please

**Review by shunting314 — APPROVED** [review]
pad-as-cat can be applied to any general padding?
